pallene/reaching_definitions.lua

```diff
@@ -0,0 +1,497 @@
+    except for the control flow ones (Nop, Seq, Return, Break, Loop, If and For).
+
+    Values aren't nodes.
+
+--]]
+function CFG:new(code)
+    local new = {}
+    setmetatable(new, self)
+    self.__index = self
+    -- exit node
+    new[-1] = {cmd = false, to={},from={}}
+    -- entry node
+    new[0] = {cmd = false, to={},from={}}
+    if code then
+        local ctx = CFG.new_ctx()
+        -- set current node of new_ctx as 0
+        ctx.current_node = {0}
+        new:build(code,ctx)
+    end
+    return new
+end
+-- builds the graph and calls new_node.
+-- it receives a node to analyze and the context until now.
+-- it returns an updated context (a copy), see CFG.new_ctx
+-- TODO:(discussion) receive a callback, since this treatment is language specific
+function CFG:build(cmd,ctx)
+    local tag = cmd._tag
+    if     tag == "ir.Cmd.Nop" then
+        return ctx
+    elseif tag == "ir.Cmd.Seq" then
+        local newctx = ctx
+        for _, c in ipairs(cmd.cmds) do
+            newctx = self:build(c,newctx)
+        end
+        return newctx
+    elseif tag == "ir.Cmd.Return" then
+        -- -1 means end of the function
+        return self:new_edge({-1},ctx)
+    elseif tag == "ir.Cmd.Break" then
+        local loop = Stack.top(ctx.current_loop)
+        assert(loop,'break outside loop (should be impossible)')
+        return self:new_edge(loop, ctx)
+    elseif tag == "ir.Cmd.Loop" then
+        Stack.push(ctx.current_loop, {})
+        --print('create new loop',#ctx.current_loop)
+        local newctx = self:build(cmd.body,ctx)
+        -- at the end of the loop, gets all indexes at the top 
+        -- and makes them point to the next (TODO)
+        -- will have to change the edge logic on break
+        Stack.pop(newctx.current_loop)
+        return newctx
+    elseif tag == "ir.Cmd.If" then
+        -- condition is an ir.Value and 
+        -- will be extracted to an assignment instruction beforehand
+        local then_ = cmd.then_ and self:build(cmd.then_,ctx) or {}
+        local else_ = cmd.else_ and self:build(cmd.else_,ctx) or {}
+        return CFG.ctx_union(then_,else_)
+    elseif tag == "ir.Cmd.For" then
+        -- start,limit,step are ir.Value and 
+        -- will be extracted to assignments beforehand
+        Stack.push(ctx.current_loop, ctx.current_node)
+        local newctx = self:build(cmd.body,ctx)
+        Stack.pop(newctx.current_loop)
+        return newctx
+    elseif string.match(tag, "^ir%.Cmd%.") then
+        return self:new_node(cmd,ctx)
+    else
+        return ctx
+    end
+end
+-- returns a copy of the context
+-- a context has:
+--     current_node : set<index>,  
+--     current_loop : stack<index>
+function CFG.new_ctx(ctx)
+    local new = {}
+    new.current_node = {}
+    new.current_loop = {}
+    if ctx then
+        Set.copy(new.current_node,ctx.current_node)
+        Stack.copy(new.current_loop,ctx.current_loop)
+    end
+    return new
+end
```

Comment:
Are you sure we need to carry around a mutable context object? I suspect that we could just make these things into parameters of the recursive build function. That is how we do it in the `test` function from uninitialized.lua.